### PR TITLE
Add ebtree based TreeMap data structure

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
This adds a TreeMap, which is an ebtree based map with malloc-allocated
nodes, +with a very low overhead for the entire map.

### CODE REVIEW

File: src/ocean/util/container/TreeMap.d [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+    By default the map uses plain `eb64_node` nodes.
+    To attach values to an ebtree node, create a struct, say, `Node`, whose
+    first field is of type `eb64_node`, followed by any number of other fields,
+    and pass this struct as the `TreeMap` template parameter.
+      - Do not manipulate the first `eb64_node` field (or the plain node if not
+        passing an own struct). Reading may be useful as it contains the key.
+      - Instances of this struct handed out by `TreeMap` are malloc-allocated so
+        they are invisible to the GC.
+
+    We are using an ebtree with malloc-allocated nodes, rather than a hash map,
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/TreeMap.d]: Suggest rewording this to:
"
Compared to a standard `HashMap`, a tree-based map is useful in situations where the number of elements in the map is hard to predict. The overhead of an ebtree is very low compared to a hash map, as the overhead of a hash map is the whole bucket set array, while for an ebtree is only a little struct. 
"

File: src/ocean/util/container/TreeMap.d [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+    while for an ebtree is only a little struct.
+
+    Copyright: Copyright (c) 2016-2018 dunnhumby Germany GmbH. All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module ocean.util.container.TreeMap;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/TreeMap.d]: I wonder about putting it in `util.container.map`.

File: src/ocean/util/container/TreeMap.d [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+
+/******************************************************************************/
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/TreeMap.d]: Could get rid of this weird line, while we're at it ;)

File: src/ocean/util/container/TreeMap.d [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+
+import ocean.util.container.ebtree.c.eb64tree;
+import ocean.util.container.ebtree.c.ebtree;
+
+/*******************************************************************************
+
+    Template params:
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/TreeMap.d]: `Params:`

File: src/ocean/util/container/TreeMap.d [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+
+            bool added;
+            MyClass obj  = map.put(12345, added, new MyClass);
+            MyClass obj2 = map.getUserElement(12345);
+            map.remove(obj);
+        ---
+
+    ***************************************************************************/
+
+    static if (is(typeof(Node.user_element_with_treemap_backlink) UserElement))
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/TreeMap.d]: Suggest removing this alternative API. It's confusing and doesn't seem to really add much.

File: src/ocean/util/container/TreeMap.d [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+        public static void remove ( ref Node node )
+        {
+            static if (is(Node == eb64_node))
+                eb64_delete(&node);
+            else
+                eb64_delete(&node.tupleof[0]);
+
+            allocator.deallocate(&node);
+        }
+    }
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/TreeMap.d]: Suggest also adding `public bool remove ( ulong key )`.

**Review by gavin-norman-sociomantic — CHANGES_REQUESTED** [review]

**Comment by nemanja-boric-sociomantic** [comment]
Updated with all the points above addressed.

File: src/ocean/util/container/map/TreeMap.d [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+        else
+            alias eb64_last eb64_function;
+
+        return cast(Node*)eb64_function(&this.root);
+    }
+
+    /***********************************************************************
+
+        foreach iterator over nodes in the tree. Any tree modification is
+        permitted during iteration.
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/map/TreeMap.d:177]: This is actually a big advantage of this container, too. Worth mentioning in the relnotes / module header.

File: src/ocean/util/container/map/TreeMap.d [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+            }
+            else
+            {
+                free(node);
+            }
+        }
+    }
+}
+
+/*******************************************************************************
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/map/TreeMap.d]: Annoying ocean comment: a usage example and unittest would be nice! :grimacing:

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/map/TreeMap.d]: Sigh... ;-)

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/map/TreeMap.d]: I think we cut out the wrong API. As this stands, this map is just a map of `id`s, since AFAICS `eb64_node` (to which we get the pointer from `Put`) can't hold any user defined value. We need the "other" API that supports any arbitrary data type, as long as they have `eb_node` as a first element.

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/map/TreeMap.d]: This having said, I'm not convinced anymore this is a good fit for Ocean :-/.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the API based around user defined type and few extras (instead of `getBoundary` we now have `first` and `last` methods, `opApply` gives both key and the value, etc). Example unittest is added as well.

File: src/ocean/util/container/map/TreeMap.d [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+
+    EBtree based map.
+
+    By default the map uses plain `eb64_node` nodes.
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/map/TreeMap.d]: The module doc needs to be updated.

File: src/ocean/util/container/map/TreeMap.d [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+    }
+
+    /***********************************************************************
+
+        foreach iterator over nodes in the tree. Any tree modification is
+        permitted during iteration.
+
+    ***********************************************************************/
+
+    public int opApply ( int delegate ( ref ulong key, ref T* node ) dg )
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/map/TreeMap.d]: Suggest changing `node` -> `value`.

File: src/ocean/util/container/map/TreeMap.d [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+
+        return stop;
+    }
+
+
+    /***********************************************************************
+
+        Removes node from the map and deallocates it.
+
+        DO NOT USE node from this point!
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/map/TreeMap.d]: I guess this comment no longer applies.

File: src/ocean/util/container/map/TreeMap.d [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+            eb_node = next;
+        }
+
+        return stop;
+    }
+
+
+    /***********************************************************************
+
+        Removes node from the map and deallocates it.
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/map/TreeMap.d]: Needs updating.

File: src/ocean/util/container/map/TreeMap.d [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+    }
+}
+
+version (UnitTest)
+{
+    import ocean.core.Test;
+}
+
+///
+unittest {
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/map/TreeMap.d]: Unconventional style.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Yeah, this looks much better now!

File: src/ocean/util/container/map/TreeMap.d (hunk 10 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/map/TreeMap.d]: Thanks, I've missed this one.

File: src/ocean/util/container/map/TreeMap.d (hunk 11 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/map/TreeMap.d]: Yeah. It should be "DO NOT USE any previous pointers to the accompanying value from this point!" or something similar.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with all comments addressed.

File: src/ocean/util/container/map/TreeMap.d [hunk 14]
```diff
@@ -0,0 +1,484 @@
+/*******************************************************************************
+
+    EBtree based map.
+
+    Compared to a standard HashMap, a tree-based map is useful in situations
+    where the number of elements in the map is hard to predict. The overhead
+    of an ebtree is very low compared to a hash map, as the overhead of a hash
+    map is the whole bucket set array, while for an ebtree is only a little
+    struct.
+
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/map/TreeMap.d]: Should also mention the modification / iteration thing here.

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/map/TreeMap.d]: Sorry I could swear I added it here. Perhaps it was lost in refactoring.

Comment (gavin-norman-sociomantic) [reply src/ocean/util/container/map/TreeMap.d]: You added it to the release notes.

**Comment by nemanja-boric-sociomantic** [comment]
Updated.

File: src/ocean/util/container/map/TreeMap.d (hunk 14 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/map/TreeMap.d]: You are right, I've checked `git reflog` and it was never committed.

**Review by gavin-norman-sociomantic — APPROVED** [review]
